### ISSUE TITLE
Need to produce document in PDF/A-3

### ISSUE THREAD

**Issue by pp717** [QuestPDF/QuestPDF#754]
Many organizations in my country allow documents to be sent as PDF files.
But it must be PDF/A-3 only because the data must be processed further.
Now in the official document, we can choose to make it as PDF/A-2b or not.
 
I would like you to urgently update to PDFA-3.

**Comment by MarcinZiabek** [comment]
What feature exactly do you need from the PDF/A-3 standard? Would you please elaborate a little bit more about your use-case?

**Comment by Rafael94** [comment]
@MarcinZiabek 

In Germany, electronic invoices in the B2B sector will be mandatory from 1 January 2025. These must be available in the ZUGFeRD 2.0 specification. This is based on the EU Directive 2014/55/EU (https://ec.europa.eu/digital-building-blocks/sites/display/DIGITAL/Types+of+digital+invoices)

Invoices must be sent as PDF/A-3 (https://en.wikipedia.org/wiki/PDF/A#PDF/A-3) with embedded XML based on the ZUGFeRD specification. 

Unfortunately, there is no alternative to this format, which is why all companies must adapt their systems.

This feature is also likely to affect the following issues:
#802 
#815

**Comment by MarcinZiabek** [comment]
@Rafael94 Thank you so much for sharing this information ❤️ 

I am excited to say that I am currently investigating support for PDF/A-3, embedding files, PDF transformations and securing files with passwords. These features are complex and may require time to introduce, but they are becoming one of the highest priorities for upcoming months.

**Comment by Rafael94** [comment]
Thanks for the feedback

**Comment by TrAnn3l** [comment]
Hey, any progress on this?

**Comment by MarcinZiabek** [comment]
Yes 😄 

This will become the highest priority once the QuestPDF Companion App is finished, which is quite close to completion.

Of course, the initial effort is to modify the structure of the PDF file so it conforms with the PDF/A-3 standard without adding new features. The next phase would focus on introducing new features.

Are you aware of any additional requirements related to EU Directive 2014/55/EU that we should care about?

**Comment by Rafael94** [comment]
I am not aware of any other requirements. It is only necessary to attach an XML file. I assume that the customer's merchandise management program extracts the XML from the PDF and processes it further.

**Comment by AlexJaouen** [comment]
As in Germany, in France, electronic invoices in the B2B sector will be mandatory from september 2026. PDF/A-3 with EN 16931 norm will be largely used in the next mounths.

**Comment by KronerPa** [comment]
Hey,

I saw that the latest version supports pdf a3 metadata right now,.

I like this library and im curious, if the extension to attachment xml files will available in the next couple of weeks.

**Comment by Rafael94** [comment]
It can already be tested in version 2024.12.0-rc0.

Instructions are also available: https://www.questpdf.com/concepts/document-operations.html

In my opinion, it is just not practical to use. The generated PDF and the attachments must be stored as a file in the file system.  The new class loads the PDF and the attachments from the file system and merges them. It is then saved back to the file system. If you then want to send it by e-mail, it has to be loaded again.

If the files are not to be saved in the file system, they must be deleted again after processing.

**Comment by Mr-MooMoo** [comment]
@Rafael94 Yeah,t hat would absolutely come in handy, as in most use cases you need to transfer the result PDF over some wire anyway.

I have my source PDF + source XML as MemoryStream, and I need the result as Stream (or byte[]), too. But currently, i need to write a Temp-File and read it back several times.

**Comment by Mr-MooMoo** [comment]
> Im pro MemoryStream too, but i wanna check out the resulting pdfs. Did anyone of you had problems loading the qpdf.dll? As the error message suggested i tried forcing the architectures to run qpdf unsuccessfully. I tried it for plain new console apps for .net 8 and .net 9 on win11.

Yes, same here. See my [Issue here](https://github.com/QuestPDF/QuestPDF/issues/1057).

**Comment by MarcinZiabek** [comment]
Thank you for sharing your feedback regarding the newly introduced `DocumentOperation` feature. I truly appreciate your involvement! 😄

I completely agree that, in many cases, it’s more practical to work with streams or byte arrays rather than files. At the moment, extending the API to efficiently use memory is challenging due to the current QuestPDF <-> qpdf integration. However, it’s certainly possible, and I’ll add this task to the backlog. As I anticipate further changes and enhancements built on top of the qpdf library - dedicating time to this improvement makes sense.

In the meantime, I’ll experiment with alternative ways to support byte arrays and enhance the API. One idea is to implement an internal wrapper that still relies on files but provides a more convenient interface for developers. While streams are also an option, they may offer less value since they would need to be rewindable, effectively existing as in-memory byte arrays.

Could you please share your thoughts on the current API? What options or changes would you suggest to improve the developer experience?

**Comment by KronerPa** [comment]
Im fine with the overall developer experience. Its a good idea to wrap the current lack of MemoryStream support as it could be swapped out later without changing the exposed calls.
Im just a bit worried about the platform troubles qpdf seems to have.

**Comment by MarcinZiabek** [comment]
> Im just a bit worried about the platform troubles qpdf seems to have.

Would you like to share what problems are concerning?

**Comment by KronerPa** [comment]
Its about the issue from @Mr-MooMoo mentioned above.

**Comment by Rafael94** [comment]
I have an idea how to make it easier to use. It hides the current problem, but should the integration of qpdf be better, not much would have to be changed.

The `IDocumentContainer` interface could be extended with the `AddAttachment` function. This function expects a stream or byte[] + filename. When `GeneratePdf` is called, the PDF is created as before. It could then be checked if there are any attachments. If so, a `System.IO.Path.GetTempPath()` temporary file is created for the PDF and for each attachment. Then 
```
DocumentOperation.LoadFile(tempPDF)
    .AddAttachment(new DocumentAttachment
    {
        FilePath = attachment[i].tempFile,
        Description = attachment[i].description,
        MimeType = attachment[i].MimeType
    })
    .Save(tempPDF)
```
The created PDF is re-imported. The temporary files are deleted and the generated PDF is returned.

So at least not every developer would have to implement it himself.

@MarcinZiabek

**Comment by faisalbwn** [comment]
I added the attachment as guided but couldn't verify it as PDF/A3-b by checking on following URL:
https://demo.verapdf.org/

**Comment by KronerPa** [comment]
I did evaluate my pdfs with a different online tooling for conformance. I guess it was pdf2go but i cant remember.
I tried the verapdf tooling and could successfully extract the xml file aswell.

I guess we need more specific information to evaluate the problem.

**Comment by faisalbwn** [comment]
**I am adding attachment by using following code:**

`// Add xml attachment
DocumentOperation.LoadFile(pdfTmpFilePath)
				.AddAttachment(new DocumentAttachment
				{
					FilePath = xmlTmpFilePath,
					//AttachmentName = $"{viewModel.XmlAttachmentName}.xml",
					Description = "Invoice XML data",
					MimeType = "application/xml",
					Relationship = DocumentAttachmentRelationship.Alternative,
					CreationDate = DateTime.UtcNow,
					ModificationDate = DateTime.UtcNow
				}).Save(pdfFilePath.FilePath);`

**But when i verify it from Pdf2Go, i got the following errors:**

The validation for file GeneratedFile.pdf against PDF/A-3B validation profile was not successful, your file broke the following rules:

    The Catalog dictionary of a conforming file shall contain the Metadata key whose value is a metadata stream as defined in ISO 32000-1:2008, 14.3.2
    DeviceRGB shall only be used if a device independent DefaultRGB colour space has been set when the DeviceRGB colour space is used, or if the file has a PDF/A OutputIntent that contains an RGB destination profile

**Comment by KronerPa** [comment]
Sounds for me like the rgb aren't embedded in the pdf.
Did you verify to set the pdfA flag in the document setting?
https://www.questpdf.com/concepts/document-settings.html#document-settings

**Comment by faisalbwn** [comment]
i found the issue, actually i was setting the PdfA to "true" in IDocument Constructor and it was not reflecting. 
I override the settings like below and it started working:

`public DocumentSettings GetSettings()
{
    // Return document settings
    return new DocumentSettings()
    {
        PdfA = true
    };
}`

**Comment by MarcinZiabek** [comment]
Following the suggestion, QuestPDF now produces PDF/A-3b conformant documents starting with the 2024.10.3 release.

As @faisalbwn pointed out, using the proper configuration is essential. You can find more details here: [Document Settings](https://www.questpdf.com/concepts/document-settings.html). Hope this helps!

Since there has been no recent activity, I’m closing this issue. If you have a similar request in the future, feel free to open a new GitHub issue. Thanks for your feedback! ❤️
